src/main/scala/is/hail/methods/SplitMulti.scala

```diff
@@ -1,158 +1,213 @@
+    require(!(removeMoving && verifyLeftAligned))
+
+    ur.set(rv)
+    val v = ur.getAs[Variant](1)
+
+    var isLeftAligned = true
+    if (prevLocus != null && prevLocus == v.locus)
+      isLeftAligned = false
+    var splitVariants = v.altAlleles.iterator.zipWithIndex
       .filter(keepStar || !_._1.isStar)
-      .map { case (aa, aai) => (Variant(v.contig, v.start, v.ref, Array(aa)).minRep, aai + 1) }
-      .filter { case (sv, _) => f(sv) }
+      .map { case (aa, aai) =>
+        val splitv = Variant(v.contig, v.start, v.ref, Array(aa))
+        val minsplitv = splitv.minRep
+
+        if (splitv.locus != minsplitv.locus)
+          isLeftAligned = false
+
+        (minsplitv, aai + 1)
+      }
       .toArray
 
     if (splitVariants.isEmpty)
       return Iterator()
 
-    val splitGenotypeBuilders = splitVariants.map { case (sv, _) => new GenotypeBuilder(sv.nAlleles) }
-    val splitGenotypeStreamBuilders = splitVariants.map { case (sv, _) => new ArrayBuilder[Genotype]() }
-
-    for (g <- it) {
-      val gadsum = Genotype.ad(g).map(gadx => (gadx, gadx.sum))
-
-      // svj corresponds to the ith allele of v
-      for (((svj, i), j) <- splitVariants.iterator.zipWithIndex) {
-        val gb = splitGenotypeBuilders(j)
-        gb.clear()
-
-        if (g == null)
-          gb.setMissing()
-        else {
-          Genotype.gt(g).foreach { ggtx =>
-            val gtx = splitGT(ggtx, i)
-            gb.setGT(gtx)
-
-            val p = Genotype.gtPair(ggtx)
-            if (gtx != p.nNonRefAlleles)
-              gb.setFakeRef()
-          }
-
-          gadsum.foreach { case (gadx, sum) =>
-            // what bcftools does
-            // Array(gadx(0), gadx(i))
-            gb.setAD(Array(sum - gadx(i), gadx(i)))
-          }
-
-          Genotype.dp(g).foreach { dpx => gb.setDP(dpx) }
-
-          if (propagateGQ)
-            Genotype.gq(g).foreach { gqx => gb.setGQ(gqx) }
-
-          Genotype.pl(g).foreach { gplx =>
-            val plx = gplx.iterator.zipWithIndex
-              .map { case (p, k) => (splitGT(k, i), p) }
-              .reduceByKeyToArray(3, Int.MaxValue)(_ min _)
-            gb.setPX(plx)
-
-            if (!propagateGQ) {
-              val gq = Genotype.gqFromPL(plx)
-              gb.setGQ(gq)
-            }
-          }
-        }
-
-        splitGenotypeStreamBuilders(j) += gb.result()
-      }
+    if (isLeftAligned) {
+      if (removeLeftAligned)
+        return Iterator()
+    } else {
+      if (removeMoving)
+        return Iterator()
+      else if (verifyLeftAligned)
+        fatal("found non-left aligned variant: $v")
     }
 
-    splitVariants.iterator
-      .zip(splitGenotypeStreamBuilders.iterator)
-      .map { case ((v, ind), gsb) =>
-        (v, (insertSplitAnnots(va, ind, true), gsb.result()))
-      }
-  }
+    val va = ur.get(2)
 
-  def splitNumber(str: String): String =
-    if (str == "A" || str == "R" || str == "G")
-      "."
-    else str
+    if (v.isBiallelic) {
```

Comment:
you mentioned on Thursday that the fact that gAnnotator is ignored in the biallelic might pose a problem if people were to write their own expressions for the row-splitting. Does that make sense to change though if the point of the expressions is to provide a way to split the rows and biallelic rows don't need to be split?

Reply:
That's correct.  I removed the special case handling of biallelics.  I think we should look at the performance once all the region value stuff is working nicely and then decide if it is necessary to special case biallelics again.